sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/action/QueryChangeStreamAction.java

```diff
@@ -187,6 +193,7 @@ public ProcessContinuation run(
               changeStreamRecordMapper.toChangeStreamRecords(
                   updatedPartition, resultSet.getCurrentRowAsStruct(), resultSet.getMetadata());
 
+          double recordsBytes = 0D;
```

Comment:
Could we refactor so that the logic of computing the recordBytes is encapsulated in the throughputEstimator. I was thinking of something in the lines of:

```
throughputEstimator.update(record);
```
Within this call we would calculate the bytes (`record.toString().getBytes(StandardCharsets.UTF_8).length;`) and accumulate.

Reply:
Yes, I did this way originally. But I found that all records here come from a single row so they have the same commit timestamp. That's why I aggregate the bytes of all records first and then update once instead of calling update for each record. 

Do you think I should change it or keep the current way?

Reply:
I will change it for simplicity. If we detect any performance issue, then we can update it.

Reply:
Just thought a bit again. I'd prefer to keep the current way as it is more easy to test. Otherwise, I need to construct a record which has exactly a certain number of bytes, e.g., 10 bytes?

Reply:
Also for simplicity reason, the current throughput estimator is more like a pure function which can be easily tested. I'd not prefer to import a dependency to it.

Reply:
I will update it to: 

```
throughputEstimator.update(record.getRecordTimestamp(), record.toString().getBytes(StandardCharsets.UTF_8).length);
```

Reply:
Thanks!